### PR TITLE
Extended library with ignoreList parameter

### PR DESCRIPTION
**Pull request by JeffreyArts**
For a brainfart I'm working on, I would like to iterate through a directory to search for *.js files. This lib is great for using that. But I wanted too exclude certain directories (like the 'node_modules' dir for example). This is why I extended the library with an option 'ignoreList'. This parameter can contain an array with files or directories which should be excluded from the result json object.

If you like this feature you could add it to the main branch, otherwise just reject it. No hard feelings ;-)

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -70,16 +70,110 @@ To:
...
 
 ```javascript
 var dirTree = require('directory-tree');
-var filteredTree = dirTree.directoryTree('/some/path', ['.jpg', '.png']);
+
+// Get full directory tree
+var tree = dirTree.directoryTree('/photos');
+
+// Only get the structure which contains *.jpg or *.png files.
+var filteredTree = dirTree.directoryTree('/photos',  {'fileTypes': ['.jpg', '.png']} );
```
Comment (mihneadb) [inline README.md]: This doesn't match the code (there's no `filetypes` in the code, there is `extensions`).

File: lib/directory-tree.js [hunk 2]
```diff
@@ -1,38 +1,77 @@
...
+                });
+
+
+                if (item.children.length == 0) {
+                    return null;
+                }
             }
         }
 
         return item;
```
Comment (mihneadb) [inline lib/directory-tree.js]: If `ignore` is `true`, `item` is `undefined`.

File: lib/directory-tree.js [hunk 3]
```diff
@@ -1,38 +1,77 @@
...
+ * @param  {string} string  The haystack which you want to search through
+ * @param  {array} array    The needles which you are looking for
+ * @return {boolean}        True if a needle is found, otherwise false.
+ */
+searchSubstring = function(string, array){
+    var found = false;
+    array.forEach( function(value, key){
+        if (string.indexOf( value ) > 0 ) {
+            found = true;
+            return true;
```
Comment (mihneadb) [inline lib/directory-tree.js]: I think this return statement is meaningless. This loop should definitely have a way to early return.

File: lib/directory-tree.js [hunk 4]
```diff
@@ -1,38 +1,77 @@
...
+
+function directoryTree(basepath, options) {
+
+    var _directoryTree = function (name, options) {
+
+
+        var ignore = false;
+        if ( typeof options.ignoreList === "object") {
+            // // Check if the name (file | directory) is on the ignore list
+            if ( searchSubstring(name, options.ignoreList) ) {
```
Comment (mihneadb) [inline lib/directory-tree.js]: You can simply early return instead of do unnecessary complex nesting.

**Comment by mihneadb** [comment]
Hi, thanks for the PR. I've added some comments on your code. I'm not sure about this. Maybe if you improve the implementation and make it cleaner we could merge it.

File: lib/directory-tree.js (hunk 3 above)
Comment (JeffreyArts) [inline lib/directory-tree.js]: What do you mean by this? The reason why it returns anything is because it stops the foreach when found something =]

Comment (mihneadb) [inline lib/directory-tree.js]:
```
Note: There is no way to stop or break a forEach() loop. The solution is to use Array.prototype.every() or Array.prototype.some(). See example below.
```
From: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/forEach

**Pull request closed without merging**